File: typescript/libs/workspace/adapters/system/src/lib/session/Session.ts
import { IChangeSet, InvokeOptions, IObject, IPullResult, IPushResult, IResult, ISession, ISessionServices, Method, Procedure, Pull } from '@allors/workspace/domain/system';
import { Workspace } from '../workspace/Workspace';
import { SessionOriginState } from './originstate/SessionOriginState';
import { Strategy } from './Strategy';
import { ChangeSetTracker } from './trackers/ChangSetTracker';
import { PushToDatabaseTracker } from './trackers/PushToDatabaseTracker';
import { PushToWorkspaceTracker } from './trackers/PushToWorkspaceTracker';
import { ChangeSet } from './ChangeSet';
import { enumerate } from '../collections/Range';
import { AssociationType, Class, Composite, Origin, RoleType } from '@allors/workspace/meta/system';

export function isNewId(id: number): boolean {
  return id < 0;
}

export abstract class Session implements ISession {
  changeSetTracker: ChangeSetTracker;

  pushToDatabaseTracker: PushToDatabaseTracker;

  pushToWorkspaceTracker: PushToWorkspaceTracker;

  sessionOriginState: SessionOriginState;

  strategyByWorkspaceId: Map<number, Strategy>;

  private strategiesByClass: Map<Class, Set<Strategy>>;

  constructor(public workspace: Workspace, public services: ISessionServices) {
    this.strategyByWorkspaceId = new Map();
    this.strategiesByClass = new Map();
    this.sessionOriginState = new SessionOriginState();
    this.changeSetTracker = new ChangeSetTracker();
    this.pushToDatabaseTracker = new PushToDatabaseTracker();
    this.pushToWorkspaceTracker = new PushToWorkspaceTracker();
  }

  abstract Create(cls: Composite): IObject;

  abstract invoke(method: Method | Method[], options?: InvokeOptions): Promise<IResult>;

  abstract call(procedure: Procedure, ...pulls: Pull[]): Promise<IPullResult>;

  abstract pull(pulls: Pull[]): Promise<IPullResult>;

  abstract push(): Promise<IResult>;

  getOne<T>(id: number): T {
    return (this.getStrategy(id)?.object as unknown) as T;
  }

  getMany<T>(ids: number[]): T[] {
    return (ids.map((v) => this.getStrategy(v)).filter((v) => v != null) as unknown) as T[];
  }

  getAll<T extends IObject>(objectType: Composite): T[] {
    const all: T[] = [];

    for (const cls of objectType.classes) {
      switch (cls.origin) {
        case Origin.Workspace:
          if (this.workspace.workspaceIdsByWorkspaceClass.has(cls)) {
            const ids = this.workspace.workspaceIdsByWorkspaceClass.get(cls);
            for (const id of ids) {
              if (this.strategyByWorkspaceId.has(id)) {
                const strategy = this.strategyByWorkspaceId.get(id);
                all.push(strategy.object as T);
              } else {
                const strategy = this.instantiateWorkspaceStrategy(id);
                all.push(strategy.object as T);
              }
            }
          }

          break;
      }
    }

    return all;
  }

  checkpoint(): IChangeSet {
    const changeSet = new ChangeSet(this, this.changeSetTracker.Created, this.changeSetTracker.Instantiated);
    if (this.changeSetTracker.DatabaseOriginStates != null) {
      for (const databaseOriginState of this.changeSetTracker.DatabaseOriginStates) {
        databaseOriginState.Checkpoint(changeSet);
      }
    }

    if (this.changeSetTracker.WorkspaceOriginStates != null) {
      for (const workspaceOriginState of this.changeSetTracker.WorkspaceOriginStates) {
        workspaceOriginState.Checkpoint(changeSet);
      }
    }

    this.sessionOriginState.Checkpoint(changeSet);
    this.changeSetTracker.Created = null;
    this.changeSetTracker.Instantiated = null;
    this.changeSetTracker.DatabaseOriginStates = null;
    this.changeSetTracker.WorkspaceOriginStates = null;

    return changeSet;
  }

  public getStrategy(id: number): Strategy {
    if (id == 0) {
      return;
    }

    if (this.strategyByWorkspaceId.has(id)) {
      return this.strategyByWorkspaceId.get(id);
    }

    if (isNewId(id)) {
      this.instantiateWorkspaceStrategy(id);
    }

    return null;
  }

  public getRole(association: Strategy, roleType: RoleType): any {
    const role = this.sessionOriginState.Get(association.id, roleType);
    if (roleType.objectType.isUnit) {
      return role;
    }

    if (roleType.isOne) {
      return this.getOne<IObject>(role);
    }

    if (role !== null) {
      const roles = [];
      for (const v of enumerate(role)) {
        roles.push(this.getOne<IObject>(v));
      }
    } else {
      return [];
    }
  }

  public getCompositeAssociation<T extends IObject>(role: number, associationType: AssociationType): T {
    const roleType = associationType.roleType;
    for (const association of this.getForAssociation(associationType.objectType as Composite)) {
      if (association.canRead(roleType)) {
        if (association.isCompositeAssociationForRole(roleType, role)) {
          return association.object as T;
        }
      }
    }
  }

  public getCompositesAssociation<T extends IObject>(role: number, associationType: AssociationType): T[] {
    const roleType = associationType.roleType;

    const associations: T[] = [];

    for (const association of this.getForAssociation(associationType.objectType as Composite)) {
      if (!association.canRead(roleType)) {
        // TODO: Warning!!! continue If
      }

      if (association.isCompositesAssociationForRole(roleType, role)) {
        associations.push(association.object as T);
      }
    }

    return associations;
  }

  protected addStrategy(strategy: Strategy) {
    this.strategyByWorkspaceId.set(strategy.id, strategy);

    let strategies = this.strategiesByClass.get(strategy.cls);
    if (strategies == null) {
      strategies = new Set();
      this.strategiesByClass.set(strategy.cls, strategies);
    }

    strategies.add(strategy);
  }

  protected removeStrategy(strategy: Strategy) {
    this.strategyByWorkspaceId.delete(strategy.id);

    const strategies = this.strategiesByClass.get(strategy.cls);
    if (strategies != null) {
      strategies.delete(strategy);
    }
  }

  protected pushToWorkspace(result: IPushResult): IPushResult {
    if (this.pushToWorkspaceTracker.Created != null) {
      for (const strategy of this.pushToWorkspaceTracker.Created) {
        strategy.WorkspaceOriginState.Push();
      }
    }

    if (this.pushToWorkspaceTracker.Changed != null) {
      for (const state of this.pushToWorkspaceTracker.Changed) {
        if (this.pushToWorkspaceTracker.Created?.has(state.strategy) == true) {
          continue;
        }

        state.Push();
      }
    }

    this.pushToWorkspaceTracker.Created = null;
    this.pushToWorkspaceTracker.Changed = null;
    return result;
  }

  protected onDatabasePushResponseNew(workspaceId: number, databaseId: number) {
    const strategy = this.strategyByWorkspaceId[workspaceId];
    this.pushToDatabaseTracker.Created.delete(strategy);
    this.removeStrategy(strategy);
    strategy.OnDatabasePushNewId(databaseId);
    this.addStrategy(strategy);
    this.onDatabasePushResponse(strategy);
  }

  protected onDatabasePushResponse(strategy: Strategy) {
    const databaseRecord = this.workspace.database.onPushResponse(strategy.cls, strategy.id);
    strategy.onDatabasePushResponse(databaseRecord);
  }

  private getForAssociation(objectType: Composite): Strategy[] {
    const classes = objectType.classes;

    const strategies: Strategy[] = [];

    for (const [_, strategy] of this.strategyByWorkspaceId) {
      if (classes.has(strategy.cls)) {
        strategies.push(strategy);
      }
    }

    return strategies;
  }

  abstract instantiateDatabaseStrategy(id: number): Strategy;

  abstract instantiateWorkspaceStrategy(id: number): Strategy;
}

File: dotnet/System/workspace/csharp/adapters/allors.workspace.adapters/session/Session.cs
// <copyright file="RemoteSession.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Threading.Tasks;
    using Data;
    using Meta;
    using Ranges;

    public abstract class Session : ISession
    {
        private readonly Dictionary<IClass, ISet<Strategy>> strategiesByClass;

        protected Session(Workspace workspace, ISessionServices sessionServices)
        {
            this.Workspace = workspace;
            this.Services = sessionServices;

            this.StrategyByWorkspaceId = new Dictionary<long, Strategy>();
            this.strategiesByClass = new Dictionary<IClass, ISet<Strategy>>();
            this.SessionOriginState = new SessionOriginState(this.Workspace.Ranges);

            this.ChangeSetTracker = new ChangeSetTracker();
            this.PushToDatabaseTracker = new PushToDatabaseTracker();
            this.PushToWorkspaceTracker = new PushToWorkspaceTracker();

            this.Services.OnInit(this);
        }

        public ISessionServices Services { get; }

        IWorkspace ISession.Workspace => this.Workspace;
        public Workspace Workspace { get; }

        public ChangeSetTracker ChangeSetTracker { get; }

        public PushToDatabaseTracker PushToDatabaseTracker { get; }

        public PushToWorkspaceTracker PushToWorkspaceTracker { get; }

        public SessionOriginState SessionOriginState { get; }

        protected Dictionary<long, Strategy> StrategyByWorkspaceId { get; }

        public abstract T Create<T>(IClass @class) where T : class, IObject;

        public abstract Task<IInvokeResult> Invoke(Method method, InvokeOptions options = null);

        public abstract Task<IInvokeResult> Invoke(Method[] methods, InvokeOptions options = null);

        public abstract Task<IPullResult> Call(Procedure procedure, params Pull[] pulls);

        public abstract Task<IPullResult> Pull(params Pull[] pulls);

        public abstract Task<IPushResult> Push();

        public T Create<T>() where T : class, IObject => this.Create<T>((IClass)this.Workspace.DatabaseConnection.Configuration.ObjectFactory.GetObjectType<T>());

        public T GetOne<T>(IObject @object) where T : IObject => this.GetOne<T>(@object.Id);

        public T GetOne<T>(T @object) where T : IObject => this.GetOne<T>(@object.Id);

        public T GetOne<T>(long? id) where T : IObject => id.HasValue ? this.GetOne<T>((long)id) : default;

        public T GetOne<T>(long id) where T : IObject => (T)this.GetStrategy(id)?.Object;

        public T GetOne<T>(string idAsString) where T : IObject => long.TryParse(idAsString, out var id) ? (T)this.GetStrategy(id)?.Object : default;

        public IEnumerable<T> GetMany<T>(IEnumerable<IObject> objects) where T : IObject => objects.Select(this.GetOne<T>);

        public IEnumerable<T> GetMany<T>(IEnumerable<T> objects) where T : IObject => objects.Select(this.GetOne);

        public IEnumerable<T> GetMany<T>(IEnumerable<long> ids) where T : IObject => ids.Select(this.GetOne<T>);

        public IEnumerable<T> GetMany<T>(IEnumerable<string> ids) where T : IObject => this.GetMany<T>(ids.Select(
            v =>
            {
                long.TryParse(v, out var id);
                return id;
            }));

        public IEnumerable<T> GetAll<T>() where T : IObject
        {
            var objectType = (IComposite)this.Workspace.DatabaseConnection.Configuration.ObjectFactory.GetObjectType<T>();
            return this.GetAll<T>(objectType);
        }

        public IEnumerable<T> GetAll<T>(IComposite objectType) where T : IObject
        {
            foreach (var @class in objectType.Classes)
            {
                switch (@class.Origin)
                {
                    case Origin.Workspace:
                        if (this.Workspace.WorkspaceIdsByWorkspaceClass.TryGetValue(@class, out var ids))
                        {
                            foreach (var id in ids)
                            {
                                if (this.StrategyByWorkspaceId.TryGetValue(id, out var strategy))
                                {
                                    yield return (T)strategy.Object;
                                }
                                else
                                {
                                    strategy = this.InstantiateWorkspaceStrategy(id);
                                    yield return (T)strategy.Object;
                                }
                            }
                        }

                        break;
                    case Origin.Database:
                    case Origin.Session:
                        if (this.strategiesByClass.TryGetValue(@class, out var strategies))
                        {
                            foreach (var strategy in strategies)
                            {
                                yield return (T)strategy.Object;
                            }
                        }

                        break;
                    default:
                        throw new ArgumentOutOfRangeException($"Origin {@class.Origin}");
                }
            }
        }

        public IChangeSet Checkpoint()
        {
            var changeSet = new ChangeSet(this, this.ChangeSetTracker.Created, this.ChangeSetTracker.Instantiated);

            if (this.ChangeSetTracker.DatabaseOriginStates != null)
            {
                foreach (var databaseOriginState in this.ChangeSetTracker.DatabaseOriginStates)
                {
                    databaseOriginState.Checkpoint(changeSet);
                }
            }

            if (this.ChangeSetTracker.WorkspaceOriginStates != null)
            {
                foreach (var workspaceOriginState in this.ChangeSetTracker.WorkspaceOriginStates)
                {
                    workspaceOriginState.Checkpoint(changeSet);
                }
            }

            this.SessionOriginState.Checkpoint(changeSet);

            this.ChangeSetTracker.Created = null;
            this.ChangeSetTracker.Instantiated = null;
            this.ChangeSetTracker.DatabaseOriginStates = null;
            this.ChangeSetTracker.WorkspaceOriginStates = null;

            return changeSet;
        }

        public Strategy GetStrategy(long id)
        {
            if (id == 0)
            {
                return null;
            }

            if (this.StrategyByWorkspaceId.TryGetValue(id, out var sessionStrategy))
            {
                return sessionStrategy;
            }

            return IsNewId(id) ? this.InstantiateWorkspaceStrategy(id) : null;
        }

        public object GetUnitRole(Strategy association, IRoleType roleType) => this.SessionOriginState.GetUnitRole(association.Id, roleType);

        public long? GetCompositeRole(Strategy association, IRoleType roleType) => this.SessionOriginState.GetCompositeRole(association.Id, roleType);

        public IRange GetCompositesRole(Strategy association, IRoleType roleType) => this.SessionOriginState.GetCompositesRole(association.Id, roleType);

        public Strategy GetCompositeAssociation(long role, IAssociationType associationType)
        {
            var roleType = associationType.RoleType;

            if (roleType.ObjectType.IsUnit)
            {
                throw new ArgumentException("AssociationType should not be for a Unit", nameof(associationType));
            }

            foreach (var association in this.StrategiesForClass(associationType.ObjectType))
            {
                if (!association.CanRead(roleType))
                {
                    continue;
                }

                if (association.IsCompositeAssociationForRole(roleType, role))
                {
                    return association;
                }
            }

            return null;
        }

        public IEnumerable<Strategy> GetCompositesAssociation(long role, IAssociationType associationType)
        {
            var roleType = associationType.RoleType;

            if (roleType.ObjectType.IsUnit)
            {
                throw new ArgumentException("AssociationType should not be for a Unit", nameof(associationType));
            }

            foreach (var association in this.StrategiesForClass(associationType.ObjectType))
            {
                if (!association.CanRead(roleType))
                {
                    continue;
                }

                if (association.IsCompositesAssociationForRole(roleType, role))
                {
                    yield return association;
                }
            }
        }

        protected abstract Strategy InstantiateWorkspaceStrategy(long id);

        protected void AddStrategy(Strategy strategy)
        {
            this.StrategyByWorkspaceId.Add(strategy.Id, strategy);

            var @class = strategy.Class;
            if (!this.strategiesByClass.TryGetValue(@class, out var strategies))
            {
                this.strategiesByClass[@class] = new HashSet<Strategy> { strategy };
            }
            else
            {
                strategies.Add(strategy);
            }
        }

        protected IPushResult PushToWorkspace(IPushResult result)
        {
            if (this.PushToWorkspaceTracker.Created != null)
            {
                foreach (var strategy in this.PushToWorkspaceTracker.Created)
                {
                    strategy.WorkspaceOriginState.Push();
                }
            }

            if (this.PushToWorkspaceTracker.Changed != null)
            {
                foreach (var state in this.PushToWorkspaceTracker.Changed)
                {
                    if (this.PushToWorkspaceTracker.Created?.Contains(state.Strategy) == true)
                    {
                        continue;
                    }

                    state.Push();
                }
            }

            this.PushToWorkspaceTracker.Created = null;
            this.PushToWorkspaceTracker.Changed = null;

            return result;
        }

        protected void OnDatabasePushResponseNew(long workspaceId, long databaseId)
        {
            var strategy = this.StrategyByWorkspaceId[workspaceId];
            this.PushToDatabaseTracker.Created.Remove(strategy);

            this.RemoveStrategy(strategy);
            strategy.OnDatabasePushNewId(databaseId);
            this.AddStrategy(strategy);

            this.OnDatabasePushResponse(strategy);

        }

        protected void OnDatabasePushResponse(Strategy strategy)
        {
            var databaseRecord = this.Workspace.DatabaseConnection.OnPushResponse(strategy.Class, strategy.Id);
            strategy.OnDatabasePushResponse(databaseRecord);
        }

        internal static bool IsNewId(long id) => id < 0;

        private void RemoveStrategy(Strategy strategy)
        {
            this.StrategyByWorkspaceId.Remove(strategy.Id);

            var @class = strategy.Class;
            if (!this.strategiesByClass.TryGetValue(@class, out var strategies))
            {
                return;
            }

            strategies.Remove(strategy);
        }

        private IEnumerable<Strategy> StrategiesForClass(IComposite objectType)
        {
            var classes = new HashSet<IClass>(objectType.Classes);
            return this.StrategyByWorkspaceId.Where(v => classes.Contains(v.Value.Class)).Select(v => v.Value);
        }
    }
}
